### PR TITLE
macOS port

### PR DESCRIPTION
**Pull request by rejsmont**
I managed to get py-libzfs to compile on macOS with OpenZFS on OSX. Here are the changes required to get this working. This includes one dirty hack to fix the compilation error due to header conflict. If you have a better idea on how to solve this, I am happy to test and discuss it!

### CODE REVIEW

**Comment by rejsmont** [comment]
Here is the configure log:

```
./configure                                                                                                                 3.69 
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed

-----------------------------------
   RUNNING TESTS ON zfs.h HEADER
-----------------------------------
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/local/bin/ggrep
checking for egrep... /usr/local/bin/ggrep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for sys/types.h... (cached) yes
checking for sys/fs/zfs.h... yes
checking for ZFS_MAX_DATASET_NAME_LEN defined in sys/fs/zfs.h... yes
configure: Checking signature of zfs_prop_valid_for_type in zfs header
configure: 3 PARAMS
configure: Checking existence of VDEV_AUX_ASHIFT_TOO_BIG enum in zfs header
configure: Not found
configure: Checking existence of pss_pass_scrub_pause & pss_pass_scrub_spent_paused struct member
configure: OK
configure: Checking existence of pss_pass_issued struct member
configure: OK
configure: Checking existence of zpool_errata_t
configure: OK
configure: Checking typedef pool_scrub_cmt_t in zfs header
configure: OK
configure: Verifying existence of zpool_rewind_policy_t
configure: NOT FOUND
configure: Verifying existence of zpool_load_policy_t
configure: OK

----------------------------------------
   RUNNING TESTS ON libzutil.h HEADER
----------------------------------------
checking for libzutil.h... no
configure: NOT FOUND
configure: Checking existence of zpool_read_label in libzutil.h header
configure: NOT FOUND
configure: Checking existence of zpool_search_import in libzutil.h header
configure: NOT FOUND

--------------------------------------
   RUNNING TESTS ON libzfs.h HEADER
--------------------------------------
checking for libzfs.h... yes
configure: Checking attribute name for verbose in sendflags_t struct
configure: Attribute named as verbose in sendflags_t struct
configure: Checking function signature zvol_volsize_to_reservation
configure: 2 PARAMS
configure: Checking existence of zpool_search_import in libzfs.h header
configure: OK
configure: Checking zfs_iter_snapshots parameters
configure: 6 params
configure: Checking existence of zpool_read_label with 2 params in libzfs.h header
configure: NOT FOUND
configure: Checking existence of zpool_read_label with 3 params in libzfs.h header
configure: OK
configure: Checking existence of vs_*_ashift in fs/zfs.h header
configure: NOT FOUND
configure: Checking function signature of zpool_get_status
configure: 3 PARAMS
configure: Checking function signature of zpool_scan
configure: 3 params
configure: Checking function signature of zfs_send_one
configure: 4 params
configure: Checking EZFS_SCRUB_PAUSED presence
configure: OK
configure: Checking ZPOOL_STATUS_NON_NATIVE_ASHIFT presence
configure: NOT FOUND
configure: Checking typedef sendflags_t in libzfs header
configure: OK
configure: Checking LZC_SEND_FLAG_EMBED_DATA presence
configure: OK
configure: Checking params for zfs_receive
configure: 6 params
configure: Checking existence of zfs_send_resume and zfs_send_resume_token_to_nvlist
configure: OK
checking for ZFS_MAXNAMELEN defined in libzfs.h... no
configure: Checking existence of renameflags_t in libzfs.h
configure: NOT FOUND
checking for ZPOOL_MAXNAMELEN defined in libzfs.h... no
configure: Checking for resumable member in recvflags_t struct
configure: OK
configure: Checking for compress member in sendflags_t struct
configure: OK

-------------------------------------------
   RUNNING TESTS ON libzfs_core.h HEADER
-------------------------------------------
checking for libzfs_core.h... yes
configure: Verifying lzc_bookmark existence
configure: OK
configure: Verifying existence and params of lzc_send_space
configure: 4 params

-----------------------------------------
   RUNNING TESTS ON zfs_ioctl.h HEADER
-----------------------------------------
configure: Verifying existence of zfs_ioctl.h
checking for zfs_ioctl.h... no
configure: NOT FOUND

--------------------------------------
   RUNNING TESTS ON nvpair.h HEADER
--------------------------------------
checking for sys/nvpair.h... yes
configure: Nvpair.h header under sys

-------------------------
   GENERATING CONFIG.H
-------------------------
configure: GENERATED pxd/config.pxi

--------------------------
   GENERATING config.py
--------------------------
configure: GENERATED config.py
configure: creating ./config.status
config.status: creating pxd/config.pxi
config.status: creating config.py
config.status: creating Makefile
```

**Comment by rejsmont** [comment]
Here is the make log:

```
make                                                                                                                        4.23 
/usr/local/bin/python3.7 setup.py build
running build
running build_ext
cythoning libzfs.pyx to libzfs.c
warning: ./pxd/nvpair.pxd:56:3: 'DATA_TYPE_UINT8_ARRAY' redeclared
warning: libzfs.pyx:576:24: Non-trivial type declarators in shared declaration (e.g. mix of pointers and values). Each pointer declaration should be on its own line.
warning: libzfs.pyx:576:31: Non-trivial type declarators in shared declaration (e.g. mix of pointers and values). Each pointer declaration should be on its own line.
warning: libzfs.pyx:577:29: Non-trivial type declarators in shared declaration (e.g. mix of pointers and values). Each pointer declaration should be on its own line.
warning: libzfs.pyx:577:35: Non-trivial type declarators in shared declaration (e.g. mix of pointers and values). Each pointer declaration should be on its own line.
building 'libzfs' extension
creating build
creating build/temp.macosx-10.15-x86_64-3.7
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Tk.framework/Versions/8.5/Headers -I/usr/local/include -I/usr/local/opt/openssl@1.1/include -I/usr/local/opt/sqlite/include -I/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/include/python3.7m -c libzfs.c -o build/temp.macosx-10.15-x86_64-3.7/libzfs.o -g -O2 -I../zfs//include/ -I../zfs//lib/libspl/include/ -I./include/ -D_MACHINE_ENDIAN_H_ -DHAVE_ISSETUGID -include ../zfs//include/sys/zfs_mount.h -Werror=implicit-function-declaration
libzfs.c:15880:13: warning: code will never be executed [-Wunreachable-code]
  __pyx_r = 0;
            ^
libzfs.c:19935:52: warning: passing 'const char **' to parameter of type 'char **' discards qualifiers in nested pointer types
      [-Wincompatible-pointer-types-discards-qualifiers]
      __pyx_v_ret = zfs_is_mounted(__pyx_v_handle, (&__pyx_v_mntpt));
                                                   ^~~~~~~~~~~~~~~~
../zfs//include/libzfs.h:834:56: note: passing argument to parameter here
extern boolean_t zfs_is_mounted(zfs_handle_t *, char **);
                                                       ^
libzfs.c:46826:13: warning: code will never be executed [-Wunreachable-code]
  __pyx_r = 0;
            ^
libzfs.c:62180:85: warning: passing 'const char *' to parameter of type 'char *' discards qualifiers
      [-Wincompatible-pointer-types-discards-qualifiers]
        __pyx_v_ret = zfs_destroy_snaps(__pyx_v_self->__pyx_base.__pyx_base.handle, __pyx_v_c_name, __pyx_v_defer_deletion);
                                                                                    ^~~~~~~~~~~~~~
../zfs//include/libzfs.h:678:52: note: passing argument to parameter here
extern int zfs_destroy_snaps(zfs_handle_t *, char *, boolean_t);
                                                   ^
libzfs.c:71800:5: warning: code will never be executed [-Wunreachable-code]
    PyErr_Format(PyExc_ValueError, "Assignment to slice of wrong length, expected %" CYTHON_FORMAT_SSIZE_T "d, got %" CYTHON_FORMAT_...
    ^~~~~~~~~~~~
libzfs.c:71787:5: warning: code will never be executed [-Wunreachable-code]
    PyErr_Format(PyExc_ValueError, "Assignment to slice of wrong length, expected %" CYTHON_FORMAT_SSIZE_T "d, got %" CYTHON_FORMAT_...
    ^~~~~~~~~~~~
libzfs.c:71765:5: warning: code will never be executed [-Wunreachable-code]
    PyErr_Format(PyExc_ValueError, "Assignment to slice of wrong length, expected %" CYTHON_FORMAT_SSIZE_T "d, got %" CYTHON_FORMAT_...
    ^~~~~~~~~~~~
libzfs.c:72260:5: warning: code will never be executed [-Wunreachable-code]
    PyErr_Format(PyExc_ValueError, "Assignment to slice of wrong length, expected %" CYTHON_FORMAT_SSIZE_T "d, got %" CYTHON_FORMAT_...
    ^~~~~~~~~~~~
libzfs.c:72247:5: warning: code will never be executed [-Wunreachable-code]
    PyErr_Format(PyExc_ValueError, "Assignment to slice of wrong length, expected %" CYTHON_FORMAT_SSIZE_T "d, got %" CYTHON_FORMAT_...
    ^~~~~~~~~~~~
libzfs.c:72225:5: warning: code will never be executed [-Wunreachable-code]
    PyErr_Format(PyExc_ValueError, "Assignment to slice of wrong length, expected %" CYTHON_FORMAT_SSIZE_T "d, got %" CYTHON_FORMAT_...
    ^~~~~~~~~~~~
libzfs.c:90511:26: warning: code will never be executed [-Wunreachable-code]
                module = PyImport_ImportModuleLevelObject(
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11 warnings generated.
creating build/lib.macosx-10.15-x86_64-3.7
clang -bundle -undefined dynamic_lookup -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk build/temp.macosx-10.15-x86_64-3.7/libzfs.o -L/usr/local/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/sqlite/lib -lnvpair -lzfs -lzfs_core -luutil -o build/lib.macosx-10.15-x86_64-3.7/libzfs.cpython-37m-darwin.so -L/usr/local/lib
```

**Comment by william-gr** [comment]
Thanks! We definitely dont have a macOS to try this out but we will just make sure this doesnt break other builds before merging this.

**Comment by rejsmont** [comment]
> Thanks! We definitely dont have a macOS to try this out but we will just make sure this doesnt break other builds before merging this.

Perhaps @lundman from @openzfsonosx could test this on a mac.

**Comment by lundman** [comment]
I could replicate the same output as author:

```
# sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15
BuildVersion:	19A602

>>> import libzfs
>>> pools = list(libzfs.ZFS().pools)
>>> print(pools)
[<libzfs.ZFSPool name 'rpool' guid '7057657535755181343'>, <libzfs.ZFSPool name 'tank' guid '4884774378527453884'>]
```

Should it not be able to find `zfs_ioctl.h` ? Is the dirty header hack referring to the `assert` work around?

**Comment by rejsmont** [comment]
Thanks @lundman for taking some time to test!

> Should it not be able to find `zfs_ioctl.h` ?

It totally should! I do not know why, but in `configure.ac`, the `zfs_ioctl.h` was referenced as `zfs_ioctl.h` instead of `sys/zfs_ioctl.h`. I checked the FreeBSD ZFS source tree, and indeed it is `sys/zfs_ioctl.h`. Corrected this in this commit.

```
-----------------------------------------
   RUNNING TESTS ON zfs_ioctl.h HEADER
-----------------------------------------
configure: Verifying existence of zfs_ioctl.h
checking for sys/zfs_ioctl.h... yes
configure: OK
```

> Is the dirty header hack referring to the `assert` work around?

Affirm! The CFLAGS set early on by cython seem to break the header chain and the `__assert` macro required in `libspl/include/assert.h` gets lost. I played around with re-including header search paths, but could not find a working configuration. Hence the dirty hack. If someone knows how to fix the chain without this, I would be very happy to include a proper solution.

**Comment by rejsmont** [comment]
Just to add some testing, import search works nicely. That was the only functional change as on macOS import search requires `kthread` initialization:

```
Python 3.7.5 (default, Nov  1 2019, 02:16:23)
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import libzfs
>>> pools = list(libzfs.ZFS().find_import())
>>> print(pools)
[<libzfs.ZFSImportablePool name 'Storage' guid '3173139757797647876'>]
>>>
```

File: configure.ac
Comment (sonicaj) [inline configure.ac]: `HAVE_O3X` won't be set as `0/False` when this is executed which will result in FreeBSD/linux builds failing. We need to fix this.

Comment (sonicaj) [reply configure.ac]: We need to make sure in case it's not macos we are building this on, this is set to 0 ( like we have done for various cases below )

File: pxd/config.pxi.in
Comment (sonicaj) [inline pxd/config.pxi.in]: I think it's not a good approach to make this platform specific - we should perhaps be checking for existence of `thread_init` and `thread_fini` in the configure.ac and defining them here. As these are only macos specific, they won't exist on other platforms and not executed during libzfs usage

**Review by sonicaj — CHANGES_REQUESTED** [review]
Thank you @rejsmont, I have a couple of notes

**Comment by rejsmont** [comment]
@sonicaj  Thank you for your suggestions. I think they are fixed with this commit.

File: configure.ac
Comment (rejsmont) [reply configure.ac]: This is now fixed in 9657ab4

File: pxd/config.pxi.in
Comment (rejsmont) [reply pxd/config.pxi.in]: This is now fixed in 9657ab4

**Comment by rejsmont** [comment]
Ping @sonicaj @william-gr

**Review by sonicaj — COMMENTED** [review]
@rejsmont it looks good to me. I'll be testing linux/freebsd builds today. However meanwhile can you please update `configure.ac` to use tabs and not a mix of tabs/spaces please ? That's how we have done it right now for that file and it would be great if we could adhere to that. Thank you

File: configure.ac
Comment (sonicaj) [inline configure.ac]: Unnecessary tab added ?

File: configure.ac
Comment (sonicaj) [inline configure.ac]: Here as well ?

File: configure.ac
Comment (sonicaj) [inline configure.ac]: As above

File: configure.ac
Comment (sonicaj) [inline configure.ac]: As above

File: configure.ac
Comment (sonicaj) [inline configure.ac]: As above

File: configure.ac
Comment (sonicaj) [inline configure.ac]: As above

**Review by sonicaj — COMMENTED** [review]
Just some minor nits again please wrt tabs/spaces. Can you please have them addressed ? Thank you

**Comment by rejsmont** [comment]
> Just some minor nits again please wrt tabs/spaces. Can you please have them addressed ? Thank you

Fixed! Sorry about that...

**Review by sonicaj — APPROVED** [review]
Looks good to me. Thank you @rejsmont

**Comment by sonicaj** [comment]
@rejsmont just FYI: We are making some branch changes to the repository. So while they are in progress ( should be resolved this week hopefully ), I'll hold off merging this to master. Once they are done, we will go ahead with merging. Thanks again for the contribution !

**Comment by rejsmont** [comment]
Awesome! Was pleasure!

**Comment by sonicaj** [comment]
@rejsmont we have shifted the branch and this is ready to go in. However we have some conflicts, can you please rebase your master branch on top of py-libzfs master branch ?

Thank you

**Comment by rejsmont** [comment]
Done!

File: pxd/libzfs.pxd
Comment (sonicaj) [inline pxd/libzfs.pxd]: A minor nit, the last line should have a new line character at the end. It was there but got lost in the rebase.

**Comment by sonicaj** [comment]
@rejsmont you added some more bits in your last commit :P Please have a peek at the diff of last commit. You have also added `config.py` and changed indentation of a couple of lines as well. Can you fix those please ? Thank you

File: configure.ac
Comment (sonicaj) [inline configure.ac]: I just noticed `zof=true`, `zof` means `zfs_on_freebsd` which this is clearly not. Can you please confirm if this is intentional ? If yes, please add a new variable perhaps as this will cause confusion. And I am not sure if this was there before 🤔 Maybe got in recent commits ?

File: configure.ac
Comment (sonicaj) [inline configure.ac:67]: Do you think `with_o3x` is necessary ?
If it's just the path to header files, why can't we use `header_prefix` ?

File: configure.ac
Comment (sonicaj) [inline configure.ac:90]: `header_prefix` is being used here instead of `with_o3x`, is this intentional ?

File: configure.ac
Comment (sonicaj) [inline configure.ac]: With the current layout, when can `zof` be `false` with `with_o3x` being not empty while building on macos ?
I suspect that won't happen and I am not sure I follow the motivation behind this conditional then. Can you please elaborate ?

**Review by sonicaj — COMMENTED** [review]
@rejsmont I was going through the PR again to be sure I hadn't missed anything and I found a couple more queries. Can you please let me know what you think about them ?
Your contribution is much appreciated. Thank you

File: configure.ac
Comment (rejsmont) [reply configure.ac]: Yeah, probably should rename this to `o3x` and use it wherever `zof` was used.

File: configure.ac
Comment (rejsmont) [reply configure.ac:67]: Standard o3x installation does not include headers. You need to pull the source to get the headers. Hence with_o3x allows one to specify the location of o3x source tree.

File: configure.ac
Comment (rejsmont) [reply configure.ac:90]: Contrary to the headers, libraries are installed in the standard location.

File: configure.ac
Comment (rejsmont) [reply configure.ac]: unlike FreeBSD, macos build requires `sys/zfs_ioctl.h`. This conditional will change once I switch from `zof` to `o3x` variable.

**Comment by rejsmont** [comment]
@sonicaj I changed the variable used to configure macos checks as requested.

**Comment by sonicaj** [comment]
Thank you @rejsmont for your contribution.

**Pull request merged**
